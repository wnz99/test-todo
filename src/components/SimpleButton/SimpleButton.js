import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

import { Intent, Icon } from '../../const/ui';

const buttonIntent = {
  [Intent.PRIMARY]: 'btn-primary',
  [Intent.DANGER]: 'btn-danger',
};

const SimpleButton = ({
  icon,
  intent,
  children,
  className,
  onClick,
  active,
  responsive,
  minimal,
  disabled,
}) => {
  const IconImg = Icon[icon];
  console.log('button');
  return (
    <>
      <div
        role="button"
        onClick={e => {
          e.preventDefault();
          onClick();
        }}
        onKeyDown={e => {
          if (e.key === 'Enter') {
            e.preventDefault();
            onClick();
          }
        }}
        tabIndex={0}
        className={classNames(
          'btn',
          buttonIntent[intent],
          minimal ? 'minimal' : '',
          disabled ? 'disabled' : '',
          className
        )}
      >
        <span className="btn-text">{children}</span>
        <span className="btn-img">{icon && <IconImg />}</span>
      </div>
      <style jsx>
        {`
          .btn {
            line-height: 1.5715;
            position: relative;
            display: inline-block;
            font-weight: 700;
            white-space: nowrap;
            text-align: center;
            background-image: none;
            border: 1px solid transparent;
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
            cursor: pointer;
            -webkit-transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            -webkit-user-select: none;
            -moz-user-select: none;
            -ms-user-select: none;
            user-select: none;
            -ms-touch-action: manipulation;
            touch-action: manipulation;
            height: 32px;
            padding: 2px 10px;
            font-size: 14px;
            border-radius: 2px;
            color: rgba(0, 0, 0, 0.65);
            background-color: #fff;
            border-color: #d9d9d9;
            display: flex;
            align-items: center;
            justify-content: space-between;
            opacity: 0.8;
          }
          .btn:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.5);
          }
          .btn:active {
            outline: 0;
          }
          .btn:hover {
            opacity: 1;
          }
          .btn-primary {
            border-color: #1890ff;
            text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            opacity: 0.8;
          }
          .btn-primary:hover {
            opacity: 1;
          }
          .btn-primary:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(24, 144, 255, 0.5);
          }
          .btn-danger {
            border-color: #ff7875;
            text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            opacity: 0.8;
          }
          .btn-danger:hover {
            opacity: 1;
          }
          .btn-danger:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(255, 120, 117, 0.5);
          }
          .btn-img {
            display: flex;
            align-items: center;
          }
          .minimal {
            border: none;
            box-shadow: none;
          }
        `}
      </style>
      <style jsx>
        {`
          .btn {
            min-width: ${responsive ? '0px' : '100px'};
          }
          .btn-text {
            display: ${responsive ? 'none' : 'inline'};
          }
          .btn-primary {
            color: ${active ? '#fff' : '#40a9ff'};
            background-color: ${active ? '#40a9ff' : '#fff'};
          }
          .btn-danger {
            color: ${active ? '#fff' : '#ff4d4f'};
            background-color: ${active ? '#ff4d4f' : '#fff'};
          }
          @media (--desktop) {
            .btn {
              min-width: 100px;
            }
            .btn-text {
              display: inline;
              width: 100%;
            }
            .btn-img {
              display: none;
            }
          }
        `}
      </style>
      <style jsx>
        {`
          .disabled {
            pointer-events: none;
            color: rgba(0, 0, 0, 0.25);
            background-color: #f5f5f5;
            border-color: #d9d9d9;
            text-shadow: none;
            -webkit-box-shadow: none;
            box-shadow: none;
          }
        `}
      </style>
    </>
  );
};

SimpleButton.propTypes = {
  active: PropTypes.bool,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  icon: PropTypes.string,
  intent: PropTypes.oneOf(Object.values(Intent)),
  minimal: PropTypes.bool,
  onClick: PropTypes.func,
  responsive: PropTypes.bool,
  disabled: PropTypes.bool,
};

SimpleButton.defaultProps = {
  active: false,
  className: '',
  icon: null,
  intent: Intent.NONE,
  minimal: false,
  onClick: () => {},
  responsive: true,
  disabled: false,
};

export default SimpleButton;
