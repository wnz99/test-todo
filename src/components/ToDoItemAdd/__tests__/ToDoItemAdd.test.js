/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ToDoItemAdd from '../ToDoItemAdd';
import SimpleButton from '../../SimpleButton';

const item = {
  id: 1,
  name: 'test name',
  description: 'test description',
};

const props = {
  item,
  onCancel: jest.fn(),
  onSubmit: jest.fn(),
};

describe('ToDoItemAdd component', () => {
  afterEach(() => {
    props.onCancel.mockClear();
    props.onSubmit.mockClear();
  });

  it('should render correctly', () => {
    expect(toJson(mount(<ToDoItemAdd {...props} />))).toMatchSnapshot();
  });

  it('should display form elements', () => {
    const wrapper = mount(<ToDoItemAdd {...props} />);

    expect(wrapper.find(SimpleButton).length).toEqual(2);
    expect(wrapper.find('form').length).toEqual(1);
    expect(wrapper.find('.item-add-form-label').length).toEqual(2);
    expect(wrapper.find('.item-add-from-input').length).toEqual(3);
  });

  it('should call onSubmit function when save button clicked', () => {
    const wrapper = shallow(<ToDoItemAdd {...props} />);

    const buttons = wrapper.find(SimpleButton);

    buttons.at(1).prop('onClick')();

    expect(props.onSubmit).toHaveBeenCalledTimes(1);
  });

  it('should call onCancel function when cancel button clicked', () => {
    const wrapper = shallow(<ToDoItemAdd {...props} />);

    const buttons = wrapper.find(SimpleButton);

    buttons.at(0).prop('onClick')();

    expect(props.onCancel).toHaveBeenCalledTimes(1);
  });
});
