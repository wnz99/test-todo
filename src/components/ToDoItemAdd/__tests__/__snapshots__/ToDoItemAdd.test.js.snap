// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItemAdd component should render correctly 1`] = `
<ToDoItemAdd
  item={
    Object {
      "description": "test description",
      "id": 1,
      "name": "test name",
    }
  }
  onCancel={[MockFunction]}
  onSubmit={[MockFunction]}
>
  <div
    className="item-add-container"
  >
    <header>
      <h3>
        Enter item details
      </h3>
    </header>
    <form>
      <div
        className="item-add-form-entry"
      >
        <label
          className="item-add-form-label"
          htmlFor="taskName"
        >
          Name
        </label>
        <input
          className="item-add-from-input"
          defaultValue="test name"
          id="taskName"
          name="taskName"
          type="text"
        />
      </div>
      <div
        className="item-add-form-entry"
      >
        <label
          className="item-add-form-label"
          htmlFor="taskDescription"
        >
          Description
        </label>
        <textarea
          className="item-add-from-input"
          defaultValue="test description"
          id="taskDescription"
          name="taskDescription"
        />
      </div>
      <input
        className="item-add-from-input"
        name="taskId"
        type="hidden"
        value={1}
      />
      <div
        className="item-add-form-controls"
      >
        <SimpleButton
          active={false}
          className=""
          disabled={false}
          icon={null}
          intent="none"
          minimal={false}
          onClick={[MockFunction]}
          responsive={false}
        >
          <div
            className="btn"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="btn-text"
            >
              Cancel
            </span>
            <span
              className="btn-img"
            />
          </div>
          <style>
            
          .btn {
            line-height: 1.5715;
            position: relative;
            display: inline-block;
            font-weight: 700;
            white-space: nowrap;
            text-align: center;
            background-image: none;
            border: 1px solid transparent;
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
            cursor: pointer;
            -webkit-transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            -webkit-user-select: none;
            -moz-user-select: none;
            -ms-user-select: none;
            user-select: none;
            -ms-touch-action: manipulation;
            touch-action: manipulation;
            height: 32px;
            padding: 2px 10px;
            font-size: 14px;
            border-radius: 2px;
            color: rgba(0, 0, 0, 0.65);
            background-color: #fff;
            border-color: #d9d9d9;
            display: flex;
            align-items: center;
            justify-content: center;
            opacity: 0.8;
          }
          .btn:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.5);
          }
          .btn:active {
            outline: 0;
          }
          .btn:hover {
            opacity: 1;
          }
          .btn-primary {
            border-color: #1890ff;
            text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            opacity: 0.8;
          }
          .btn-primary:hover {
            opacity: 1;
          }
          .btn-primary:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(24, 144, 255, 0.5);
          }
          .btn-danger {
            border-color: #ff7875;
            text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            opacity: 0.8;
          }
          .btn-danger:hover {
            opacity: 1;
          }
          .btn-danger:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(255, 120, 117, 0.5);
          }
          .btn-img {
            display: flex;
            align-items: center;
          }
          .minimal {
            border: none;
            box-shadow: none;
          }
        
          </style>
          <style>
            
          .btn {
            min-width: 100px;
          }
          .btn-text {
            display: inline;
          }
          .btn-primary {
            color: #40a9ff;
            background-color: #fff;
          }
          .btn-danger {
            color: #ff4d4f;
            background-color: #fff;
          }
          @media (--desktop) {
            .btn {
              min-width: 100px;
            }
            .btn-text {
              display: inline;
              width: 100%;
            }
            .btn-img {
              display: none;
            }
          }
        
          </style>
          <style>
            
          .disabled {
            pointer-events: none;
            color: rgba(0, 0, 0, 0.25);
            background-color: #f5f5f5;
            border-color: #d9d9d9;
            text-shadow: none;
            -webkit-box-shadow: none;
            box-shadow: none;
          }
        
          </style>
        </SimpleButton>
        <SimpleButton
          active={false}
          className=""
          disabled={false}
          icon={null}
          intent="none"
          minimal={false}
          onClick={[Function]}
          responsive={false}
        >
          <div
            className="btn"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="btn-text"
            >
              Save
            </span>
            <span
              className="btn-img"
            />
          </div>
          <style>
            
          .btn {
            line-height: 1.5715;
            position: relative;
            display: inline-block;
            font-weight: 700;
            white-space: nowrap;
            text-align: center;
            background-image: none;
            border: 1px solid transparent;
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
            cursor: pointer;
            -webkit-transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            -webkit-user-select: none;
            -moz-user-select: none;
            -ms-user-select: none;
            user-select: none;
            -ms-touch-action: manipulation;
            touch-action: manipulation;
            height: 32px;
            padding: 2px 10px;
            font-size: 14px;
            border-radius: 2px;
            color: rgba(0, 0, 0, 0.65);
            background-color: #fff;
            border-color: #d9d9d9;
            display: flex;
            align-items: center;
            justify-content: center;
            opacity: 0.8;
          }
          .btn:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.5);
          }
          .btn:active {
            outline: 0;
          }
          .btn:hover {
            opacity: 1;
          }
          .btn-primary {
            border-color: #1890ff;
            text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            opacity: 0.8;
          }
          .btn-primary:hover {
            opacity: 1;
          }
          .btn-primary:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(24, 144, 255, 0.5);
          }
          .btn-danger {
            border-color: #ff7875;
            text-shadow: 0 -1px 0 rgba(0, 0, 0, 0.12);
            -webkit-box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            box-shadow: 0 2px 0 rgba(0, 0, 0, 0.045);
            opacity: 0.8;
          }
          .btn-danger:hover {
            opacity: 1;
          }
          .btn-danger:focus {
            outline: 0;
            box-shadow: 0 2px 0 rgba(255, 120, 117, 0.5);
          }
          .btn-img {
            display: flex;
            align-items: center;
          }
          .minimal {
            border: none;
            box-shadow: none;
          }
        
          </style>
          <style>
            
          .btn {
            min-width: 100px;
          }
          .btn-text {
            display: inline;
          }
          .btn-primary {
            color: #40a9ff;
            background-color: #fff;
          }
          .btn-danger {
            color: #ff4d4f;
            background-color: #fff;
          }
          @media (--desktop) {
            .btn {
              min-width: 100px;
            }
            .btn-text {
              display: inline;
              width: 100%;
            }
            .btn-img {
              display: none;
            }
          }
        
          </style>
          <style>
            
          .disabled {
            pointer-events: none;
            color: rgba(0, 0, 0, 0.25);
            background-color: #f5f5f5;
            border-color: #d9d9d9;
            text-shadow: none;
            -webkit-box-shadow: none;
            box-shadow: none;
          }
        
          </style>
        </SimpleButton>
      </div>
      <div
        className="item-add-form-errors"
      />
    </form>
  </div>
  <style>
    
          .h3 {
            opacity: 0.5;
          }
          textarea {
            height: 150px;
          }
          .item-add-container {
            position: relative;
            border: 1px solid #f0f0f0;
            border-radius: 2px;
            -webkit-transition: all 0.2s;
            transition: all 0.2s;
            padding: 15px;
          }
          .item-add-form-errors {
            color: #ff4d4f;
          }
          .item-add-form {
            -webkit-box-sizing: border-box;
            box-sizing: border-box;
            margin: 0;
            padding: 0;
            color: rgba(0, 0, 0, 0.65);
            font-size: 14px;
            line-height: 1.5715;
            list-style: none;
          }
          .item-add-form-entry {
            display: flex;
            flex-direction: column;
          }
          .item-add-form-controls {
            display: flex;
            justify-content: flex-end;
            padding-top: 15px;
          }
          .item-add-form-label {
            padding-top: 5px;
            padding-bottom: 5px;
          }
          .item-add-from-input {
            -webkit-box-sizing: border-box;
            box-sizing: border-box;
            margin: 0;
            padding: 0;
            font-variant: tabular-nums;
            list-style: none;
            -webkit-font-feature-settings: 'tnum';
            font-feature-settings: 'tnum';
            position: relative;
            display: inline-block;
            width: 100%;
            min-width: 0;
            padding: 4px 11px;
            color: rgba(0, 0, 0, 0.65);
            font-size: 14px;
            line-height: 1.5715;
            background-color: #fff;
            background-image: none;
            border: 1px solid #d9d9d9;
            border-radius: 2px;
            -webkit-transition: all 0.3s;
            transition: all 0.3s;
          }
        
  </style>
</ToDoItemAdd>
`;
